migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyMigrator.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.poi.xssf.usermodel.XSSFWorkbook;
+import uk.nhs.digital.ps.migrator.MigrationReport;
+import uk.nhs.digital.ps.migrator.config.ExecutionParameters;
+import uk.nhs.digital.ps.migrator.model.Property;
+import uk.nhs.digital.ps.migrator.model.nesstar.PublishingPackage;
+import uk.nhs.digital.ps.migrator.model.taxonomy.TaxonomyTerm;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * This class is responsible for providing the taxonomy definition that we are going
+ * to import into hippo and a mapping to those terms from existing P Codes.
+ */
+public class TaxonomyMigrator {
+
+    private static final String TAXONOMY_ROOT_NODE_NAME = "publication_taxonomy";
+    private static final String TAXONOMY_DEFINITION_COLUMN = "concept";
+    private static final String TAXONOMY_JCR_PATH = "/content/taxonomies/publication_taxonomy";
+    private static final String TAXONOMY_MAPPING_COLUMN_PREFIX = "Leaf";
+    private static final String P_CODE_COLUMN = "P Code";
+
+    private MigrationReport migrationReport;
+    private ExecutionParameters executionParameters;
+
+    private HashSet<String> taxonomyKeys;
+    private HashMap<String, List<String>> taxonomyMapping;
+    private TaxonomyTerm taxonomyDefinition;
+
+    public TaxonomyMigrator(MigrationReport migrationReport, ExecutionParameters executionParameters) {
+        this.migrationReport = migrationReport;
+        this.executionParameters = executionParameters;
+    }
+
+    public void init() {
+        if (taxonomyMapping == null) {
+            readTaxonomyMapping();
+        }
+
+        if (taxonomyDefinition == null) {
+            readTaxonomyDefinition();
+        }
+    }
+
+    public List<String> getTaxonomyKeys(PublishingPackage dataset) {
+        List<String> taxonomyKeys = taxonomyMapping.get(dataset.getUniqueIdentifier());
+        if (taxonomyKeys == null || taxonomyKeys.isEmpty()) {
+            migrationReport.add("No taxonomy terms were provided in the mapping for dataset: " + dataset,
+                "Dataset will be imported without any taxonomy terms.");
+            return Collections.emptyList();
+        }
+
+        Map<Boolean, List<String>> map = taxonomyKeys.stream()
+            .collect(Collectors.partitioningBy(key -> this.taxonomyKeys.contains(key)));
+
+        List<String> notAllowedKeys = map.get(false);
+        if (!notAllowedKeys.isEmpty()) {
+            migrationReport.add("Invalid taxonomy terms were present in the taxonomy mapping",
+                "Dataset: " + dataset,
+                "Taxonomy Terms: " + String.join(", ", notAllowedKeys));
+        }
+
+        return map.get(true);
+    }
+
+    public TaxonomyTerm getTaxonomyDefinition() {
+        return taxonomyDefinition;
+    }
+
+    private void readTaxonomyMapping() {
+        Path taxonomyMappingImportPath = executionParameters.getTaxonomyMappingImportPath();
+        if (taxonomyMappingImportPath == null) {
+            return;
```

Comment:
If we are to do anything in response to this being `null`, I think that throwing an exception would be better than just quietly exiting the method as it would make troubleshooting easier.

Reply:
If you are just doing taxonomy you won't have specified this parameter so you don't want to read the file in. If you are doing nesstar convert assertRequiredArgs checks for this so you would get an exception at that point.

Reply:
Got it. NVM :smile: